flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/RestartBackoffTimeStrategyFactoryLoader.java

```diff
@@ -0,0 +1,312 @@
+			if (clazz != null && RestartBackoffTimeStrategy.class.isAssignableFrom(clazz)) {
+				final Method method = clazz.getMethod(CREATE_METHOD, Configuration.class);
+				if (method != null) {
+					final Object result = method.invoke(null, configuration);
+
+					if (result != null) {
+						return (RestartBackoffTimeStrategy.Factory) result;
+					}
+				}
+			}
+		} catch (ClassNotFoundException cnfe) {
+			LOG.warn("Could not find restart strategy class {}.", restartStrategyClassName);
+		} catch (NoSuchMethodException nsme) {
+			LOG.warn("Class {} does not have static method {}.", restartStrategyClassName, CREATE_METHOD);
+		} catch (InvocationTargetException ite) {
+			LOG.warn("Cannot call static method {} from class {}.", CREATE_METHOD, restartStrategyClassName);
+		} catch (IllegalAccessException iae) {
+			LOG.warn("Illegal access while calling method {} from class {}.", CREATE_METHOD, restartStrategyClassName);
+		}
+
+		// fallback in case of an error
+		LOG.warn("Configured restart strategy {} is not valid. Fall back to no restart strategy.",
+			restartStrategyClassName);
+		return new NoRestartBackoffTimeStrategy.NoRestartBackoffTimeStrategyFactory();
+	}
+
+	/**
+	 * Decides the {@link RestartBackoffTimeStrategy} (new version restart strategy) to use and its params based on
+	 * legacy restart strategy configs. Records its class name and the params into an adapted configuration.
+	 *
+	 * <p>The decision making is as follows:
+	 * <ol>
+	 *     <li>Use the strategy related to {@link RestartStrategies.RestartStrategyConfiguration}, unless it's
+	 * null or {@link RestartStrategies.FallbackRestartStrategyConfiguration}.</li>
+	 *     <li>If the strategy is not determined, try adapting the cluster(server-side) configs of legacy strategy
+	 * for new version, unless the legacy restart strategy is not specified or not valid</li>
+	 *     <li>If the strategy is still not determined, use {@link FixedDelayRestartStrategy} if checkpointing is
+	 * enabled. Otherwise use {@link NoRestartStrategy}</li>
+	 * </ol>
+	 *
+	 * @param jobRestartStrategyConfiguration restart configuration given within the job graph
+	 * @param clusterConfiguration cluster(server-side) configuration, usually represented as jobmanager config
+	 * @param isCheckpointingEnabled if checkpointing was enabled for the job
+	 * @return adapted configuration
+	 */
+	private static Configuration getAdaptedConfiguration(
+			final RestartStrategies.RestartStrategyConfiguration jobRestartStrategyConfiguration,
+			final Configuration clusterConfiguration,
+			final boolean isCheckpointingEnabled) {
+
+		// clone a new config to not modify the existing config
+		final Configuration adaptedConfiguration = new Configuration();
+
+		// try determining restart strategy based on job level restart strategy config first
+		if (!tryAdaptingJobRestartStrategyConfiguration(jobRestartStrategyConfiguration, adaptedConfiguration)) {
```

Comment:
I think we should drop the reflective instantiation (see my other comment) and return the `RestartBackoffTimeStrategy.Factory` immediately.  Moreover, I would restructure the code a bit to avoid output arguments (`adaptedConfiguration`), which are usually discouraged:

```
		final Optional<RestartBackoffTimeStrategy.Factory> jobRestartStrategy = [...];
		final Optional<RestartBackoffTimeStrategy.Factory> clusterRestartStrategy = [...];

		return Stream.of(jobRestartStrategy, clusterRestartStrategy)
			.filter(Optional::isPresent)
			.findFirst()
			.map(Optional::get)
			.orElse(defaultRestartStrategy(checkpointingEnabled);
```

Let me know what you think.

Reply:
Agreed. We need to reorganize the code if we do not base it on the reflective way.